orca_detector/train.py

```diff
@@ -16,59 +21,51 @@
 
 RUN_TIMESTAMP = datetime.datetime.now().isoformat('-')
 
-def print_framework_versions():
-    print(f'TensorFlow version: {tf.VERSION}')
-    print(f'Keras version: {tf.keras.__version__}')
 
-def create_network():
-    """ Instantiate but don't yet fit the model."""
+def create_network(classes):
+    """ 
+    Instantiate but don't yet fit the model.
+    Create the output shape based on the classes length
+    """
 
-    sound_extractor = OrcaVGGish(load_weights=True, 
+    sound_extractor = OrcaVGGish(load_weights=True,
                                  weights='audioset',
+                                 out_dim=len(classes),
                                  pooling='avg').get_model()
 
     return sound_extractor
 
+
 def run(**params):
 
-    print_framework_versions()
-    
-    # load the dataset mappings from disk.
-    train_files, train_labels = load_dataset(orca_params.DATA_PATH, DatasetType.TRAIN)
-    validate_files, validate_labels = load_dataset(orca_params.DATA_PATH, DatasetType.VALIDATE)
+    print(f'TensorFlow version: {tf.VERSION}')
+    print(f'Keras version: {tf.keras.__version__}')
+
+    # load the dataset features from disk.
+    train_features, train_labels = load_features(
+        orca_params.DATA_PATH, DatasetType.TRAIN)
+    validate_features, validate_labels = load_features(
+        orca_params.DATA_PATH, DatasetType.VALIDATE)
+
+    # shuffle, one hot and filter labels
+    # TODO: implement shuffling
+    # TODO: implement classes filtering based on the data distributιον
     
-    training_generator = WavDataGenerator(train_files,
-                                          train_labels,
-                                          shuffle=True,
-                                          **params)
-
-    # test the generator
-    training_generator.__getitem__(0)
-
-    validation_generator = WavDataGenerator(validate_files,
-                                            validate_labels,
-                                            shuffle=True,
-                                            **params)
-
-    model = create_network()
-
-    history = model.fit_generator(generator=training_generator,
-                                  validation_data=validation_generator,
-                                  class_weight=orca_params.CLASS_WEIGHTS,
-                                  epochs=orca_params.EPOCHS,
-                                  use_multiprocessing=True,
-                                  verbose=1,
-                                  workers=1)
+    classes = set(train_labels).union(set(validate_labels))
+    encoder = create_label_encoding(list(classes))
+    train_labels = encode_labels(train_labels, encoder)
+    validate_labels = encode_labels(validate_labels, encoder)
+
+    model = create_network(classes)
+
+    history = model.fit(x=train_features,
+                        y=train_labels,
+                        validation_data=(validate_features, validate_labels),
```

Comment:
Agreed. Especially since we are changing the cardinality of the original dataset from audio samples to .98 seconds samples. There's a risk not to validate the less chatty species this way, I like our validation logic. The only thing I might change is the percentage, I think now it's 30%.

Reply:
I think it's 70/20/10.  We could maybe go to 65/25/10 but I wouldn't want to reduce our training set much further than that.